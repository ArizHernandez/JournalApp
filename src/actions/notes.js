import Swal from 'sweetalert2';

import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { Types } from '../types/types';

export const startNewNote = () => {
  return async(dispatch, getState) => {
    const {uid} = getState().auth
    const newNote = {
      title: '',
      body:  '',
      date:  new Date().getTime(),
      imageUrl: ''
    }

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activeNote(docRef.id, newNote));
    dispatch(addNewNote(docRef.id, newNote));
  }
};

export const startLoadingNotes = (uid) => {
  return async(dispatch) => {
    const notes = await loadNotes(uid);

    dispatch(setNotes(notes));
  }
}

export const startSaveNote = (note) => {
  return async(dispatch, getState) => {
    const {uid} = getState().auth;
    const noteToFirestore = {...note};
    
    noteToFirestore.date = new Date().getTime();
    delete noteToFirestore.id;

    try {
      db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)
              .then( () => process.env.NODE_ENV !== "test" && Swal.fire('Updated', 'Your note has been updated!!', 'success'))
              .catch( (err) => process.env.NODE_ENV !== "test" && Swal.fire('Upss...', err.message, 'error '));
  
      dispatch(refreshNote(note.id, note));
    } catch (error) {
      throw new Error(error);
    }
  }
}

export const startUploadFile = (file) => {
  return async(dispatch, getState) => {
    const {active: activeNote} = getState().notes;

    if(process.env.NODE_ENV !== "test"){
      Swal.fire({
        title: 'Uploading',
        text: 'Please, wait a minute...',
        allowOutsideClick: false,
        showConfirmButton: false,
        willOpen: () => {
          Swal.showLoading();
        }
      })
    }

    try {
      const fileUrl = fileUpload(file);

      activeNote.imageUrl = fileUrl;
      activeNote.date = new Date().getTime();
      dispatch(startSaveNote(activeNote));
    } catch (error) {
      throw new Error(error);
    }
    Swal.close();
  }
}

export const startDeleteNote = (id) => {
  return async(dispatch, getState) => {
    const {uid} = getState().auth;
    await db.doc(`${uid}/journal/notes/${id}`).delete();

    dispatch(noteDelete(id));
  }
}

export const addNewNote = (id, note) => ({
  type: Types.notesAddNew,
  payload: {
    id,
    ...note
  }
})

export const activeNote = (id, note) => ({
  type: Types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes) => ({
  type: Types.notesLoad,
  payload: notes
});

export const refreshNote = (id, note) => ({
  type: Types.notesUpdated,
  payload: {
    id,
    note
  }
})

export const noteDelete = (id) => ({
  type: Types.notesDeleted,
  payload: id
})

export const notesLogout = () => ({
  type: Types.notesLogoutCleaning
});