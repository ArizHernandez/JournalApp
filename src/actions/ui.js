import { Types } from "../types/types";

export const setError = (msgError) => ({
  type: Types.uiSetError,
  payload: msgError
})

export const removeError = () => ({
  type: Types.uiRemoveError
})

export const startLoading = () => ({
  type: Types.uiStartLoading
})

export const finishLoading = () => ({
  type: Types.uiFinishLoading
})