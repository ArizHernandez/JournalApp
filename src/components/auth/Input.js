import React, { useState } from 'react'

export const Input = ({required = false, autoComplete ,labelText, placeholder, name, type, value, handleInputChange, expresionRegular, errorMessage}) => {
  const [isValid, setIsValid] = useState("")

  const validate = () => {
    if(expresionRegular){
      if(expresionRegular.test(value)){
        setIsValid(true);
      } else {
        setIsValid(false);
      }
    }
  }

  return (
    <div className="mb-2">
      <label className="mb-1">
        <small className={`
          ${(!isValid && isValid !== "") && "auth__input-invalid-text"}
          ${(isValid && isValid !== "") && "auth__input-valid-text"}
        `}
        >
          {labelText}
        </small>
      </label>

      <div className="auth__input-container">
        <input 
          autoComplete={autoComplete}
          required={required}
          name={name}
          type={type} 
          className={`
            ${(!isValid && isValid !== "") && "auth__input-invalid"} 
            ${(isValid && isValid !=="") && "auth__input-valid"}
            auth__input
          `}
          placeholder={placeholder}
          onKeyUp={validate}
          value={value}
          onChange={handleInputChange}
        />

        {
          (!isValid && isValid !== "")
          ? <i className="auth__input-icon auth__input-invalid-text fa fa-times-circle"></i>
          : (isValid && isValid !== "")
          && <i className="auth__input-icon auth__input-valid-text fa fa-check-circle"></i>
        }
      </div>
      {
        (!isValid && isValid !== "")
        && (<p className="auth__error-message">{errorMessage}</p>)
      }
    </div>
  )
}
