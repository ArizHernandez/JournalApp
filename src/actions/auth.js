import Swal from 'sweetalert2';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { Types } from '../types/types';
import { notesLogout } from './notes';
import { startLoading, finishLoading, setError } from './ui';


export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => dispatch(login(user.uid, user.displayName)))
            .catch((err) => Swal.fire('Error!', err.message, 'error'))
            .finally(() => dispatch(finishLoading()))
  };
}

export const startRegisterEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) => {
              await user.updateProfile({displayName: name})

              dispatch(login(user.uid, user.displayName))
            })
            .catch((err) => Swal.fire('Error!', err.message, 'error'))
            .finally(() => dispatch(finishLoading()));
  }
}

export const startGoogleLogin = () => {
  return ((dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => dispatch(login(user.uid, user.displayName)))
  });
}

export const startLogout = () => {
  return async(dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout())
    dispatch(notesLogout());
  }
}

export const login = (uid, displayName) => ({
    type: Types.login,
    payload: {
      uid, 
      displayName
    }
})

export const logout = () => ({
  type: Types.logout
});