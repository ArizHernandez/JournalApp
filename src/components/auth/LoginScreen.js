import React from 'react'

import validator from 'validator'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { removeError, setError } from '../../actions/ui'

import { useForm } from '../../hooks/useForm'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { Input } from './Input'

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const {msgError, loading} = useSelector(state => state.ui);
  const userValues = {
    email: '',
    password: ''
  };

  if(process.env.NODE_ENV === 'test'){
    userValues.email = 'test@test.com';
    userValues.password = '123456'
  }

  const [formValues, handleInputChange] = useForm({
    email: userValues.email,
    password: userValues.password
  });
  const {email, password} = formValues;

  const expresiones = {
    email: /^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/,
    password: /^.{6,100}$/
  }

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  }
  
  const handleLogin = (e) => {
    e.preventDefault();
    
    if(isFormValid()){
      dispatch(startLoginEmailPassword(email, password))
    }
  }
  
  const isFormValid = () => {

    if(!validator.isEmail(email)){
      dispatch(setError('Email is not valid'));
      return false;
    } else if(password.length < 6){
      dispatch(setError('Password is less to 6 characters'));
      return false;
    }
    
    dispatch(removeError());
    return true;
  }

  return (
    <>
      <h3 className="auth__title txt-center mb-2">Login</h3>
      <form onSubmit={handleLogin}>
        
       <Input
        required={true}
        labelText="Email:"
        name="email"
        type="email"
        placeholder="Email"
        errorMessage="Email is Required"
        expresionRegular={expresiones.email}
        handleInputChange={handleInputChange}
        value={email}
       />
        
       <Input
        labelText="Password:"
        name="password"
        type="password"
        placeholder="Password"
        expresionRegular={expresiones.password}
        handleInputChange={handleInputChange}
        value={password}
       />

        <div className="txt-center mb-2">
          <button
            className="btn btn-primary btn-block"
            type="submit"
            disabled={loading}
          >
            <span>Login</span>
            {
              (loading)
              && (<i className="fa fa-circle-notch fa-spin"></i>)
            }
          </button>
        </div>
      </form>
      
      {
        (msgError)
        && (
          <div className="auth__alert-error">
            <p>{msgError}</p>
          </div>
        )
      }
      
      <hr/>

      <div className="auth__social-network">
        <p>Login with social networks</p>  
        <div 
          className="google-btn mt-2"
          onClick={handleGoogleLogin}
        >
          <div className="google-icon-wrapper">
            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
          </div>
          <p className="btn-text">
            <b>Sign In with Google</b>
          </p>
        </div>
      </div>

      <div className="txt-end">
        <Link 
          to="/auth/register"
          className="link"
        >
          Create new account
        </Link>
      </div>
    </>
  )
}
