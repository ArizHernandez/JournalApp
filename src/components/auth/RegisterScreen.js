import React from 'react'

import validator from 'validator'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux';

import { useForm } from '../../hooks/useForm';
import { setError, removeError } from '../../actions/ui';
import { startRegisterEmailPasswordName } from '../../actions/auth';
import { Input } from './Input';

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const {msgError, loading} = useSelector( state => state.ui );
  const userValues = {
    name:'',
    email: '',
    password: '',
    password2: ''
  }

  if(process.env.NODE_ENV === 'test'){
    userValues.name = 'test';
    userValues.email = 'test@test.com';
    userValues.password = '123456';
    userValues.password2 = '123456';
  }


  const [formValues, handleInputChange] = useForm({
    name: userValues.name,
    email: userValues.email,
    password: userValues.password,
    password2: userValues.password2
  })
  const {name, email, password, password2} = formValues;

  const expresiones = {
    name: /^[a-zA-ZÀ-ÿ\s]{1,40}$/,
    email: /^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/,
    password: /^.{6,100}$/
  }

  const handleRegister = (e) => {
    e.preventDefault();

    if(isFormValid()) {
      dispatch(startRegisterEmailPasswordName(email, password, name));
    }
  }

  const isFormValid = () => {
    
    if(name.trim().length === 0){
      dispatch(setError('name is required'))
      return false;
    } else if(!validator.isEmail(email)){
      dispatch(setError('Email is required'));
      return false
    } else if(password !== password2 || password.length < 6){
      dispatch(setError('password is less to 6 characters or is not equal'));
      return false;
    }

    dispatch(removeError());
    return true;
  }

  return (
    <>
      <h3 className="auth__title txt-center mb-2">Register</h3>
      <form onSubmit={handleRegister}>

        <Input
          required={true}
          labelText="Name:"
          name="name"
          type="text"
          placeholder="Name"
          autoComplete="off"
          errorMessage="Name is not valid"
          expresionRegular={expresiones.name}
          handleInputChange={handleInputChange}
          value={name}
        />

        <Input
          required={true}
          labelText="Email:"
          name="email"
          type="email"
          placeholder="example@example.com"
          errorMessage="Email is not valid"
          expresionRegular={expresiones.email}
          handleInputChange={handleInputChange}
          value={email}
        />

        <Input
          required={true}
          labelText="Password:"
          name="password"
          type="password"
          placeholder="******"
          expresionRegular={expresiones.password}
          handleInputChange={handleInputChange}
          value={password}
        />

        <Input
          required={true}
          labelText="Confirm Password:"
          name="password2"
          type="password"  
          placeholder="******"
          expresionRegular={expresiones.password}
          handleInputChange={handleInputChange}
          value={password2}
        />

        <div className="txt-center">
          <button
            className="btn btn-primary btn-block"
            type="submit"
            disabled={loading}
          >
            <span>Sign In </span>
            {
              (loading)
              && (<i className="fa fa-circle-notch fa-spin"></i>)
            }
          </button>
        </div>
      </form>
      <div className="mt-1">
        <span>Alredy registered? </span>
        <Link 
          to="/auth/login"
          className="link"
        >
          logIn
        </Link>
      </div>
      {
        (msgError && !loading) 
        && (
          <div className="auth__alert-error mt-2">
            <p>{msgError}</p>
          </div>
        )
      }
    </>
  )
}
