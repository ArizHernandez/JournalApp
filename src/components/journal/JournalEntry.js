import React from 'react'
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({id, body, date, imageUrl, title, i}) => {
  const dispatch = useDispatch();
  const noteDate = moment(date);

  const handleSelectNote = () => {

    dispatch(activeNote(id, {body, date, imageUrl, title}));
  }

  return (
    <div 
      className={`journal__entry animate__animated animate__fadeInLeft animate__delay-0${(i*10)/2}s`}
      onClick={handleSelectNote}
    >
      
      <div className="journal__entry-body-container">
        {
          (imageUrl !== "")
          && ( 
            <div 
              className="journal__entry-picture"
            >
              <img src={imageUrl} alt={title}/>
            </div>
          ) 
        }

        <div className="journal__entry-body">
          <p className="journal__entry-title">
            {title}
          </p>
          <p className="journal__entry-content">
            {body}
          </p>
        </div>

      </div>

      <div className="journal__entry-date-box">
        <span>{noteDate.format('dddd')}</span>
        <h4>{noteDate.format('Do')}</h4>
      </div>
    </div>
  )
}
