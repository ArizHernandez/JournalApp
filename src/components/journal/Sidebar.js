import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';

import { JournalEtries } from './JournalEtries'

export const Sidebar = () => {
  const [actived, setActived] = useState(false);

  const dispatch = useDispatch();
  const {name} = useSelector(state => state.auth);

  
  const handleLogout = () => {
    dispatch(startLogout());
  }
  
  const handleAddNew = () => {
    dispatch(startNewNote());
  }

  return (
    <div className={`${actived && 'journal__sidebar-active'} journal__sidebar`}>
      <div 
        className="journal__menu-content"
        onClick={() => setActived(!actived)}
      >
        <div className={`${actived && 'journal__menu-item-closed'} journal__menu-item`}></div>
        <div className={`${actived && 'journal__menu-item-closed'} journal__menu-item`}></div>
      </div>
      
      <div className="journal__sidebar-navbar">
        <h3>
          <i className="fa fa-moon"></i>
          <span> {name}</span>
        </h3>

        <button 
          className={`btn journal__logout ${actived && "journal__logout-active"}`}
          onClick={handleLogout}
        >
          Logout
        </button>
      </div>

      <div 
        className="journal__new-entry"
        onClick={handleAddNew}
      >
        <i className="fa fa-calendar-plus fa-5x"></i>
        <p className="mt-2">New entry</p>
      </div>

      <JournalEtries />
    </div>
  )
}
