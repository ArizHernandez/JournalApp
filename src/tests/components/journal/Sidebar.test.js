import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import { mount } from 'enzyme';
import { Sidebar } from '../../../components/journal/Sidebar';
import { Provider } from 'react-redux';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';

jest.mock('../../../actions/auth', () => ({
  startLogout: jest.fn()
}))

jest.mock('../../../actions/notes', () => ({
  startNewNote: jest.fn()
}))

const middlewares = [thunk];
const muckStore = configureStore(middlewares);
const initialState = {
  auth: {},
  notes: {
    notes: [],
    active: {}
  }
}

const store = muckStore(initialState);
store.dispatch = jest.fn();

describe('Pruebas en el <Sidebar />', () => {
  const wrapper = mount(
    <Provider store={store}>
      <Sidebar />
    </Provider>
  )
  
  test('Debe cargar correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  })
  
  test('Debe de llamar el logout', () => {
    wrapper.find('.journal__logout').prop('onClick')();

    expect(startLogout).toHaveBeenCalled();
  })

  test('Debe de llamar el startNewNote', () => {
    wrapper.find('.journal__new-entry').prop('onClick')();

    expect(startNewNote).toHaveBeenCalled();
  })
})