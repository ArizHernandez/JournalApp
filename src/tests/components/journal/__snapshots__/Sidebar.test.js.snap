// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <Sidebar /> Debe cargar correctamente 1`] = `
<div
  className="false journal__sidebar"
>
  <div
    className="journal__menu-content"
    onClick={[Function]}
  >
    <div
      className="false journal__menu-item"
    />
    <div
      className="false journal__menu-item"
    />
  </div>
  <div
    className="journal__sidebar-navbar"
  >
    <h3>
      <i
        className="fa fa-moon"
      />
      <span>
         
      </span>
    </h3>
    <button
      className="btn journal__logout false"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="fa fa-calendar-plus fa-5x"
    />
    <p
      className="mt-2"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</div>
`;
